src/Compilers/Core/Portable/Emit/EditAndContinueMethodDebugInformation.cs

```diff
@@ -277,5 +300,56 @@ internal void SerializeLambdaMap(BlobBuilder writer)
         }
 
         #endregion
+
+        #region State Machine States
+
+        /// <exception cref="InvalidDataException">Invalid data.</exception>
+        private static unsafe ImmutableArray<StateMachineStateDebugInfo> UncompressStateMachineStates(ImmutableArray<byte> compressedStateMachineStates)
+        {
+            if (compressedStateMachineStates.IsDefaultOrEmpty)
+            {
+                return default;
+            }
+
+            var mapBuilder = ArrayBuilder<StateMachineStateDebugInfo>.GetInstance();
+
+            fixed (byte* ptr = &compressedStateMachineStates.ToArray()[0])
+            {
+                var blobReader = new BlobReader(ptr, compressedStateMachineStates.Length);
+
+                try
+                {
+                    int count = blobReader.ReadCompressedInteger();
+
+                    while (count > 0)
+                    {
+                        int stateNumber = blobReader.ReadCompressedSignedInteger();
+                        int syntaxOffset = blobReader.ReadCompressedInteger();
+
+                        mapBuilder.Add(new StateMachineStateDebugInfo(syntaxOffset, stateNumber));
+                        count--;
+                    }
+                }
+                catch (BadImageFormatException)
+                {
+                    throw CreateInvalidDataException(compressedStateMachineStates, blobReader.Offset);
+                }
+            }
+
+            return mapBuilder.ToImmutableAndFree();
+        }
+
+        internal void SerializeStateMachineStates(BlobBuilder writer)
+        {
+            writer.WriteCompressedInteger(StateMachineStates.Length);
+
+            foreach (StateMachineStateDebugInfo state in StateMachineStates)
+            {
+                writer.WriteCompressedSignedInteger(state.StateNumber);
```

Comment:
nit: Is there some kind of documentation for PDB format? If yes, I'd just want to confirm that the documented order is state ordinal followed by syntax offset.

Reply:
We do have some of the CDIs documented, but EnC ones are currently not.

Reply:
EnC metadata format is generally undocumented.